### PR TITLE
Added sample filter tool.

### PR DESCRIPTION
**Pull request by vruano**
Addresses issue #296.

### CODE REVIEW

**Comment by vruano** [comment]
@davidbenjamin please review.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/FilterSamples.java [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+import org.broadinstitute.hellbender.utils.tsv.TableWriter;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
+
+/**
+ * Tool to filters samples based on coverage statistics.
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/FilterSamples.java]: "filter" without 's'

File: src/main/java/org/broadinstitute/hellbender/tools/exome/FilterSamples.java [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+ * <p>
+ *     This tool can use the output from {@link CalculateCoverageStats} in order
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/FilterSamples.java]: "uses" better than "can use" if this is the only way to use.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/FilterSamples.java [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    @Argument(
+        doc = "Output rejection file.",
+        fullName = REJECTION_FILE_FULL_NAME,
+        shortName = REJECTION_FILE_SHORT_NAME,
+        optional = true
+    )
+    protected File rejectionFile;
+
+    @Argument(
+        doc = "Output file.",
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/FilterSamples.java]: expand doc a bit eg "output file of samples to keep"

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+/**
+ * Integration tests for {@link FilterSamples}.
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public class FilterSamplesIntegrationTest extends CommandLineProgramTest {
+
+    @Override
+    public String getTestedClassName() {
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: Why the override -- doesn't `getTestedClassName` by default just strip the `IntegrationTest` to get `FilterSamples`?

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        final File rejectionFile = createTempFile("reject", ".tab");
+        runCommandLine(new String[] {
+                "-" + StandardArgumentDefinitions.INPUT_SHORT_NAME, samplesFile.getPath(),
+                "-" + StandardArgumentDefinitions.OUTPUT_SHORT_NAME, outputFile.getPath(),
+                "-" + FilterSamples.MINIMUM_MEAN_COVERAGE_SHORT_NAME, String.valueOf(minMean),
+                "-" + FilterSamples.MAXIMUM_MEAN_COVERAGE_SHORT_NAME, String.valueOf(maxMean),
+                "-" + FilterTargets.REJECT_OUTPUT_FILE_SHORT_NAME, rejectionFile.getPath(),
+        });
+
+        checkLeftInTargets(leftInSamples, outputFile);
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: I detect some copy-and-paste from your last PR -- not that there's anything wrong with that, but you need `Targets` --> `Samples` in a few places!

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+                .filter(stats -> stats.mean < minMean || stats.mean > maxMean)
+                .collect(Collectors.toList());
+        final File outputFile = createTempFile("output", ".tab");
+        final File rejectionFile = createTempFile("reject", ".tab");
+        runCommandLine(new String[] {
+                "-" + StandardArgumentDefinitions.INPUT_SHORT_NAME, samplesFile.getPath(),
+                "-" + StandardArgumentDefinitions.OUTPUT_SHORT_NAME, outputFile.getPath(),
+                "-" + FilterSamples.MINIMUM_MEAN_COVERAGE_SHORT_NAME, String.valueOf(minMean),
+                "-" + FilterSamples.MAXIMUM_MEAN_COVERAGE_SHORT_NAME, String.valueOf(maxMean),
+                "-" + FilterTargets.REJECT_OUTPUT_FILE_SHORT_NAME, rejectionFile.getPath(),
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: `Samples`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        });
+
+        checkLeftInTargets(leftInSamples, outputFile);
+        checkRejectedTargets(leftOutSamples, rejectionFile, FilterSamples.SampleFilter.ExtremeCoverageMean);
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: `Samples`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    @Test(dataProvider = "targetExtremeMeanCoverageFilterData")
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: `sample`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public class FilterSamplesIntegrationTest extends CommandLineProgramTest {
+
+    @Override
+    public String getTestedClassName() {
+        return FilterSamples.class.getSimpleName();
+    }
+
+    @Test(dataProvider = "targetExtremeMeanCoverageFilterData")
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: `sample`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+                "-" + StandardArgumentDefinitions.INPUT_SHORT_NAME, samplesFile.getPath(),
+                "-" + StandardArgumentDefinitions.OUTPUT_SHORT_NAME, outputFile.getPath(),
+                "-" + FilterSamples.MINIMUM_MEAN_COVERAGE_SHORT_NAME, String.valueOf(minMean),
+                "-" + FilterSamples.MAXIMUM_MEAN_COVERAGE_SHORT_NAME, String.valueOf(maxMean),
+        });
+
+        checkLeftInTargets(leftInSamples, outputFile);
+    }
+
+    @Test(dataProvider = "targetExtremeCoverageVarianceFilterData")
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: `sample`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+                .filter(stats -> stats.variance < min || stats.variance > max)
+                .collect(Collectors.toList());
+        final File outputFile = createTempFile("output", ".tab");
+        final File rejectionFile = createTempFile("reject", ".tab");
+        runCommandLine(new String[] {
+                "-" + StandardArgumentDefinitions.INPUT_SHORT_NAME, samplesFile.getPath(),
+                "-" + StandardArgumentDefinitions.OUTPUT_SHORT_NAME, outputFile.getPath(),
+                "-" + FilterSamples.MINIMUM_COVERAGE_VARIANCE_SHORT_NAME, String.valueOf(min),
+                "-" + FilterSamples.MAXIMUM_COVERAGE_VARIANCE_SHORT_NAME, String.valueOf(max),
+                "-" + FilterTargets.REJECT_OUTPUT_FILE_SHORT_NAME, rejectionFile.getPath(),
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: `Samples`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+        });
+
+        checkLeftInTargets(leftInSamples, outputFile);
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: `Samples`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+        checkRejectedTargets(leftOutSamples, rejectionFile, FilterSamples.SampleFilter.ExtremeCoverageVariance);
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: `Samples`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    @Test(dataProvider = "targetExtremeCoverageVarianceFilterData")
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: `sample`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+                .collect(Collectors.toList());
+        final File outputFile = createTempFile("output", ".tab");
+        runCommandLine(new String[] {
+                "-" + StandardArgumentDefinitions.INPUT_SHORT_NAME, samplesFile.getPath(),
+                "-" + StandardArgumentDefinitions.OUTPUT_SHORT_NAME, outputFile.getPath(),
+                "-" + FilterSamples.MINIMUM_COVERAGE_VARIANCE_SHORT_NAME, String.valueOf(min),
+                "-" + FilterSamples.MAXIMUM_COVERAGE_VARIANCE_SHORT_NAME, String.valueOf(max),
+        });
+
+        checkLeftInTargets(leftInSamples, outputFile);
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: `Samples`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    private void checkLeftInTargets(final List<SampleCoverageStats> expectedSampleStats, final File outputFile) throws IOException {
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: `Samples`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+            protected String createRecord(final DataLine dataLine) {
+                return dataLine.get(0);
+            }
+        };
+        final List<String> outputSamples = outputReader.stream().collect(Collectors.toList());
+        final List<String> expectedSamples = expectedSampleStats.stream().map(stats -> stats.sample).collect(Collectors.toList());
+        Assert.assertEquals(outputSamples, expectedSamples);
+    }
+
+    private void checkRejectedTargets(final List<SampleCoverageStats> expectedSampleStats, final File rejectionFile,
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: `Samples`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+            for (final SampleCoverageStats stats : sampleStats) {
+                writer.writeRecord(stats);
+            }
+            return result;
+        } catch (final IOException ex) {
+            throw new UncheckedIOException(ex);
+        }
+    }
+
+    @DataProvider(name = "targetMinimumSizeFilterData")
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: This `DataProvider` is for `FilterTargets`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 19]
```diff
@@ -0,0 +1,285 @@
...
+                { targets,  -1},
+                { targets,   0},
+                { targets,   1},
+                { targets,  10},
+                { targets, 100},
+                { targets, 500},
+        };
+    }
+
+    @DataProvider(name = "targetExtremeGCFilterData")
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: Likewise, from `FilterTargets`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 20]
```diff
@@ -0,0 +1,285 @@
...
+        return new Object[][] {
+                { targets,  0},
+                { targets,  0.1},
+                { targets,  0.2},
+                { targets,  0.49},
+                { targets,  0.5},
+        };
+    }
+
+    @DataProvider(name = "targetExtremeRepeatFilterData")
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: Likewise.

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 21]
```diff
@@ -0,0 +1,285 @@
...
+                { targets,  0.2},
+                { targets,  0.49},
+                { targets,  0.5},
+                { targets,  1.0},
+                { targets,  0.9},
+                { targets,  -1.0},
+        };
+    }
+
+    @DataProvider(name = "targetExtremeCoverageMeanData")
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: Likewise.

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 22]
```diff
@@ -0,0 +1,285 @@
...
+        return new Object[][] {
+                { targets,  0, 10},
+                { targets,  10, 50},
+                { targets,  500, 600},
+                { targets,  -1.0, Double.POSITIVE_INFINITY},
+                { targets, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY},
+        };
+    }
+
+    @DataProvider(name = "targetExtremeMeanCoverageFilterData")
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: Likewise.

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java (hunk 21 above)
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: Or rename to `sample...`

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java [hunk 23]
```diff
@@ -0,0 +1,285 @@
...
+        return new Object[][] {
+                { stats,  0, 10},
+                { stats,  10, 50},
+                { stats,  90, 110},
+                { stats,  -1.0, Double.POSITIVE_INFINITY},
+                { stats, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY},
+        };
+    }
+
+    @DataProvider(name = "targetExtremeCoverageVarianceFilterData")
```
Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: `sample`

**Comment by davidbenjamin** [comment]
@vruano comments are all trivial, will take single-digit minutes.

File: src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java (hunk 4 above)
Comment (vruano) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: Refactoring may fail silently using that default behavior.

Comment (vruano) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: I mean without a compilation error. I think that in general one should avoid relaying on name convention based reflection. If it was me I rather would rather have left that method as abstract in the parent class.

Comment (davidbenjamin) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/FilterSamplesIntegrationTest.java]: Thanks for the explanation!

**Comment by vruano** [comment]
Ok thanks will fix and merge.

**Pull request merged**